security/advancedtls/advancedtls.go

```diff
@@ -0,0 +1,169 @@
+	Certificates []tls.Certificate
+	// If Certificates is nil, servers will invoke this function every time a new connection
+	// is established and servers need to present certificates to the clients, This is known as peer
+	// certificate reloading.
+	GetCertificate func(*tls.ClientHelloInfo) (*tls.Certificate, error)
+	// RootCACerts or GetRootCAs indicates the certificates trusted by the server side. If requiring
+	// mutual authentication on server side, only one of the two field must be set; otherwise these
+	// two fields are ignored.
+	RootCACerts *x509.CertPool
+	// If requiring mutual authentication and RootCACerts is nil, servers will invoke this function
+	// every time a new connection is established and servers need to check certificates sent from
+	// the clients against root CAs. This is known as trust certificate reloading.
+	GetRootCAs func(rawCerts [][]byte) (*x509.CertPool, error)
+	// If servers want clients to send certificates to prove its identities.
+	MutualAuth bool
+}
+
+func (o *ClientOptions) Config() (*tls.Config, error) {
+	if o.RootCACerts == nil && o.GetRootCAs == nil {
+		return nil, fmt.Errorf("either RootCACerts or GetRootCAs must be specified")
+	}
+	config := &tls.Config{
+		InsecureSkipVerify: true,
+		ServerName:         o.ServerNameOverride,
+	}
+	if o.Certificates != nil {
+		config.Certificates = o.Certificates
+	} else {
+		config.GetClientCertificate = o.GetClientCertificate
+	}
+	if o.RootCACerts != nil {
+		// Even if we don't need to reload root certs, we still need to build verification function for
+		// possible custom server authorization
+		o.GetRootCAs = func(rawCerts [][]byte) (*x509.CertPool, error) {
+			return o.RootCACerts, nil
+		}
+	}
+	// create a function which will reload the root cert and invoke users' VerifyPeerCertificate
+	config.VerifyPeerCertificate = buildVerifyFunc(o.GetRootCAs, o.VerifyPeerCertificate, false)
+	return config, nil
+}
+
+func (o *ServerOptions) Config() (*tls.Config, error) {
+	if o.Certificates == nil && o.GetCertificate == nil {
+		return nil, fmt.Errorf("either Certificates or GetCertificate must be specified")
+	}
+	config := &tls.Config{}
+	if o.Certificates != nil {
+		config.Certificates = o.Certificates
+	} else {
+		config.GetCertificate = o.GetCertificate
+	}
+	if !o.MutualAuth {
+		config.ClientAuth = tls.NoClientCert
+		return config, nil
+	}
+	if o.RootCACerts == nil && o.GetRootCAs == nil {
+		return nil, fmt.Errorf("server need trust certs if using mutual TLS")
+	}
+	if o.RootCACerts != nil {
+		config.RootCAs = o.RootCACerts
+		config.ClientAuth = tls.RequireAndVerifyClientCert
+		return config, nil
+	}
+	// We have to rewrite verify function if users want root certificate reloading
+	if o.GetRootCAs != nil {
+		config.ClientAuth = tls.RequireAnyClientCert
+		config.VerifyPeerCertificate = buildVerifyFunc(o.GetRootCAs, nil, true)
+	}
+	return config, nil
+}
+
+func buildVerifyFunc(
```

Comment:
Please add a comment on why we need this function -- because current tls module does not have a good support on root cert reloading. Therefore, we have to verify the peer certificate chain against the root CA certs in our verifyFunc.